File: 3.x.x/crates-utopia/cache/src/adapter/redis_cluster.rs
use std::thread;
use std::time::Duration;

use parking_lot::Mutex;
use redis::cluster::ClusterClient;
use redis::cluster::ClusterConnection;

use super::redis::envelope::Envelope;
use crate::adapter::Adapter;
use crate::error::CacheError;
use crate::feature::{clamp_retries, Retryable};
use crate::value::{is_empty_key, unix_now, CacheValue, LoadResult, SaveResult};

const CONNECTION_ERRORS: &[&str] = &[
    "went away",
    "socket",
    "read error on connection",
    "connection lost",
    "timed out",
    "timeout",
    "connection refused",
    "no connection",
    "broken pipe",
    "couldn't map cluster keyspace",
    "can't communicate with any node",
    "clusterdown",
    "is not covered by any node",
];

/// PHP `Utopia\Cache\Adapter\RedisCluster`.
pub struct RedisCluster {
    conn: Mutex<ClusterConnection>,
    seeds: Vec<String>,
    name: Option<String>,
    timeout: f64,
    read_timeout: f64,
    max_retries: i32,
    retry_delay: i32,
}

impl std::fmt::Debug for RedisCluster {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("RedisCluster")
            .field("seeds", &self.seeds)
            .field("timeout", &self.timeout)
            .field("read_timeout", &self.read_timeout)
            .finish_non_exhaustive()
    }
}

impl RedisCluster {
    pub fn new(connection: ClusterConnection, seeds: Vec<String>) -> Self {
        Self {
            conn: Mutex::new(connection),
            seeds,
            name: None,
            timeout: 1.5,
            read_timeout: 1.5,
            max_retries: 0,
            retry_delay: 1000,
        }
    }

    pub fn connect(seeds: Vec<String>) -> Result<Self, CacheError> {
        let client = ClusterClient::new(seeds.clone())?;
        let connection = client.get_connection()?;
        Ok(Self::new(connection, seeds))
    }

    #[must_use]
    pub fn get_cluster_name(&self) -> Option<&str> {
        self.name.as_deref()
    }

    #[must_use]
    pub fn get_seeds(&self) -> &[String] {
        &self.seeds
    }

    fn reconnect(&self) -> Result<(), CacheError> {
        let client = ClusterClient::new(self.seeds.clone())?;
        *self.conn.lock() = client.get_connection()?;
        Ok(())
    }

    fn execute<T, F>(&self, mut callback: F) -> Result<T, CacheError>
    where
        F: FnMut(&mut ClusterConnection) -> Result<T, CacheError>,
    {
        let max_attempts = 1 + self.max_retries;
        let mut attempts = 0;
        loop {
            let result = {
                let mut conn = self.conn.lock();
                callback(&mut conn)
            };
            match result {
                Ok(v) => return Ok(v),
                Err(err) if is_connection_error(&err.to_string()) => {
                    attempts += 1;
                    if attempts >= max_attempts {
                        return Err(err);
                    }
                    thread::sleep(Duration::from_millis(self.retry_delay.max(0) as u64));
                    let _ = self.reconnect();
                }
                Err(err) => return Err(err),
            }
        }
    }
}

fn is_connection_error(message: &str) -> bool {
    let lower = message.to_lowercase();
    CONNECTION_ERRORS.iter().any(|n| lower.contains(n))
}

impl Retryable for RedisCluster {
    fn set_max_retries(&mut self, max_retries: i32) -> &mut Self {
        self.max_retries = clamp_retries(max_retries);
        self
    }

    fn set_retry_delay(&mut self, retry_delay: i32) -> &mut Self {
        self.retry_delay = retry_delay;
        self
    }

    fn get_max_retries(&self) -> i32 {
        self.max_retries
    }

    fn get_retry_delay(&self) -> i32 {
        self.retry_delay
    }
}

impl Adapter for RedisCluster {
    fn load(&self, key: &str, ttl: i64, hash: &str) -> Result<LoadResult, CacheError> {
        let hash = if is_empty_key(hash) { key } else { hash };
        let redis_string: Option<String> =
            self.execute(|conn| Ok(redis::cmd("HGET").arg(key).arg(hash).query(conn)?))?;
        Ok(redis_string.map_or(LoadResult::Miss, |raw| {
            Envelope::decode(&raw, ttl, unix_now()).map_or(LoadResult::Miss, LoadResult::Hit)
        }))
    }

    fn save(&self, key: &str, data: &CacheValue, hash: &str) -> Result<SaveResult, CacheError> {
        if is_empty_key(key) || data.is_php_empty() {
            return Ok(SaveResult::Failed);
        }
        let hash = if is_empty_key(hash) { key } else { hash };
        let value = match Envelope::encode(data, unix_now()) {
            Ok(v) => v,
            Err(_) => return Ok(SaveResult::Failed),
        };
        match self.execute(|conn| {
            redis::cmd("HSET")
                .arg(key)
                .arg(hash)
                .arg(&value)
                .query::<i64>(conn)
                .map_err(CacheError::from)
        }) {
            Ok(_) => Ok(SaveResult::Saved(data.clone())),
            Err(_) => Ok(SaveResult::Failed),
        }
    }

    fn touch(&self, key: &str, hash: &str) -> Result<bool, CacheError> {
        let hash = if is_empty_key(hash) { key } else { hash };
        let redis_string: Option<String> =
            self.execute(|conn| Ok(redis::cmd("HGET").arg(key).arg(hash).query(conn)?))?;
        let Some(raw) = redis_string else {
            return Ok(false);
        };
        let Some(value) = Envelope::touch(&raw, unix_now()) else {
            return Ok(false);
        };
        Ok(self
            .execute(|conn| {
                redis::cmd("HSET")
                    .arg(key)
                    .arg(hash)
                    .arg(&value)
                    .query::<i64>(conn)
                    .map_err(CacheError::from)
            })
            .is_ok())
    }

    fn list(&self, key: &str) -> Result<Vec<String>, CacheError> {
        self.execute(|conn| Ok(redis::cmd("HKEYS").arg(key).query(conn)?))
    }

    fn purge(&self, key: &str, hash: &str) -> Result<bool, CacheError> {
        if !is_empty_key(hash) {
            return self.execute(|conn| {
                Ok(redis::cmd("HDEL").arg(key).arg(hash).query::<i64>(conn)? != 0)
            });
        }
        self.execute(|conn| Ok(redis::cmd("DEL").arg(key).query::<i64>(conn)? != 0))
    }

    fn flush(&self) -> Result<bool, CacheError> {
        self.execute(|conn| {
            redis::cmd("FLUSHDB").query::<String>(conn)?;
            Ok(true)
        })
    }

    fn ping(&self) -> bool {
        self.execute(|conn| {
            redis::cmd("PING").query::<String>(conn)?;
            Ok(true)
        })
        .unwrap_or(false)
    }

    fn get_size(&self) -> Result<i64, CacheError> {
        self.execute(|conn| Ok(redis::cmd("DBSIZE").query::<i64>(conn).unwrap_or(0)))
    }

    fn get_name(&self, _key: Option<&str>) -> String {
        "redis-cluster".into()
    }
}

impl RedisCluster {
    pub fn with_timeouts(mut self, timeout: f64, read_timeout: f64) -> Self {
        self.timeout = timeout;
        self.read_timeout = read_timeout;
        self
    }

    pub fn with_name(mut self, name: impl Into<String>) -> Self {
        self.name = Some(name.into());
        self
    }
}

File: src/Cache/Adapter/RedisCluster.php
<?php

namespace Utopia\Cache\Adapter;

use Exception;
use RedisCluster as Client;
use Throwable;
use Utopia\Cache\Adapter;
use Utopia\Cache\Feature\Batchable;
use Utopia\Cache\Feature\Retryable;

class RedisCluster implements Adapter, Batchable, Retryable
{
    private int $maxRetries = 0;

    private int $retryDelay = 1000;

    /**
     * @param  array<string>  $seeds
     * @param  string|array<string>|null  $auth  Password string or ['username', 'password'] array for ACL
     */
    public function __construct(protected Client $redis, protected array $seeds, protected ?string $name = null, private readonly float $timeout = 1.5, private readonly float $readTimeout = 1.5, private readonly bool $persistent = false, private readonly string|array|null $auth = null) {}

    /**
     * @param  int  $maxRetries (0-10)
     */
    public function setMaxRetries(int $maxRetries): self
    {
        $this->maxRetries = max(Retryable::MIN_RETRIES, min($maxRetries, Retryable::MAX_RETRIES));

        return $this;
    }

    /**
     * @param  int  $retryDelay time in milliseconds
     */
    public function setRetryDelay(int $retryDelay): self
    {
        $this->retryDelay = $retryDelay;

        return $this;
    }

    /**
     * @param  int  $ttl time in seconds
     * @param  string  $hash optional
     */
    public function load(string $key, int $ttl, string $hash = ''): mixed
    {
        if ($hash === '' || $hash === '0') {
            $hash = $key;
        }

        /** @var string|false */
        $redis_string = $this->execute(fn(): string => $this->redis->hGet($key, $hash));

        if ($redis_string === false || ! \is_string($redis_string)) {
            return false;
        }

        $cache = Json::decode($redis_string);

        // A purged key keeps its field until re-cached, holding a value that
        // is not an envelope.
        if (! \is_array($cache) || ! isset($cache['time'], $cache['data'])) {
            return false;
        }

        if ($cache['time'] + $ttl > time()) { // Cache is valid
            return $cache['data'];
        }

        return false;
    }

    /**
     * HMGET for an explicit field list, HGETALL when $fields is empty.
     *
     * @param  string[]  $fields
     * @param  int  $ttl time in seconds
     * @return array<string, mixed>
     */
    public function loadMany(string $key, array $fields, int $ttl): array
    {
        $now = time();

        /** @var array<string, mixed> $raw */
        $raw = (array) ($fields === []
            ? $this->execute(fn(): array => $this->redis->hGetAll($key))
            : $this->execute(fn(): array => $this->redis->hMget($key, $fields)));

        $result = [];
        foreach ($raw as $field => $value) {
            if (! \is_string($value)) {
                continue;
            }

            $cache = Json::decode($value);
            if (! \is_array($cache)) {
                continue;
            }
            if (! isset($cache['time'], $cache['data'])) {
                continue;
            }

            if ($cache['time'] + $ttl > $now) {
                $result[(string) $field] = $cache['data'];
            }
        }

        return $result;
    }

    /**
     * @param  array<int|string, mixed>|string  $data
     * @param  string  $hash optional
     * @param  int  $ttl time in seconds
     * @return bool|string|array<int|string, mixed>
     */
    public function save(string $key, array|string $data, string $hash = '', int $ttl = 0): bool|string|array
    {
        if ($key === '' || $key === '0' || empty($data)) {
            return false;
        }

        if ($hash === '' || $hash === '0') {
            $hash = $key;
        }

        try {
            $value = json_encode([
                'time' => time(),
                'data' => $data,
            ], flags: JSON_THROW_ON_ERROR);
        } catch (Throwable) {
            return false;
        }

        try {
            $this->execute(fn(): int => $this->redis->hSet($key, $hash, $value));

            if ($ttl > 0) {
                $this->execute(fn(): bool => $this->redis->expire($key, $ttl));
            }

            return $data;
        } catch (Throwable) {
            return false;
        }
    }

    /**
     * HMSET every field => value pair of $data in one round trip, then one EXPIRE.
     *
     * @param  array<string, mixed>  $data field => value
     * @param  int  $ttl time in seconds
     * @return array<string, mixed>|false
     */
    public function saveMany(string $key, array $data, int $ttl = 0): array|false
    {
        $map = [];
        foreach ($data as $field => $value) {
            try {
                $map[(string) $field] = json_encode([
                    'time' => time(),
                    'data' => $value,
                ], flags: JSON_THROW_ON_ERROR);
            } catch (Throwable) {
                return false;
            }
        }

        if ($map === []) {
            return false;
        }

        try {
            $this->execute(fn(): bool => $this->redis->hMSet($key, $map));

            if ($ttl > 0) {
                $this->execute(fn(): bool => $this->redis->expire($key, $ttl));
            }

            return $data;
        } catch (Throwable) {
            return false;
        }
    }

    /**
     * @param  string  $hash optional
     */
    public function touch(string $key, string $hash = ''): bool
    {
        if ($hash === '' || $hash === '0') {
            $hash = $key;
        }

        /** @var string|false */
        $redis_string = $this->execute(fn(): string => $this->redis->hGet($key, $hash));

        if ($redis_string === false || ! \is_string($redis_string)) {
            return false;
        }

        try {
            /** @var array{time: int, data: mixed} $cache */
            $cache = Json::decode($redis_string, JSON_THROW_ON_ERROR);
            $cache['time'] = time();
            $value = json_encode($cache, flags: JSON_THROW_ON_ERROR);
        } catch (Throwable) {
            return false;
        }

        return $this->execute(fn(): int => $this->redis->hSet($key, $hash, $value)) !== false;
    }

    /**
     * @return string[]
     */
    public function list(string $key): array
    {
        /** @var array<string> */
        $keys = (array) $this->execute(fn(): array => $this->redis->hKeys($key));

        if (empty($keys)) {
            return [];
        }

        return $keys;
    }

    /**
     * @param  string  $hash optional
     */
    public function purge(string $key, string $hash = ''): bool
    {
        if ($hash !== '' && $hash !== '0') {
            return (bool) $this->execute(fn(): int => $this->redis->hdel($key, $hash));
        }

        return (bool) $this->execute(fn(): int => $this->redis->del($key));
    }

    public function flush(): bool
    {
        return (bool) $this->execute(function (): true {
            /** @var array<string> $masters */
            $masters = $this->redis->_masters();
            foreach ($masters as $master) {
                $this->redis->flushDB($master);
            }

            return true;
        });
    }

    public function ping(): bool
    {
        try {
            return (bool) $this->execute(function (): true {
                foreach ($this->redis->_masters() as $master) {
                    $this->redis->ping($master);
                }

                return true;
            });
        } catch (Exception) {
            return false;
        }
    }

    /**
     * Returning total number of keys
     */
    public function getSize(): int
    {
        $size = $this->execute(function () {
            $size = 0;
            foreach ($this->redis->_masters() as $master) {
                $size += $this->redis->dbSize($master);
            }

            return $size;
        });

        if ($size === false || ! is_numeric($size)) {
            return 0;
        }

        return (int) $size;
    }

    public function getMaxRetries(): int
    {
        return $this->maxRetries;
    }

    public function getRetryDelay(): int
    {
        return $this->retryDelay;
    }

    public function getClusterName(): ?string
    {
        return $this->name;
    }

    /**
     * @return array<string>
     */
    public function getSeeds(): array
    {
        return $this->seeds;
    }

    /**
     * Execute a Redis command with retry logic
     *
     *
     * @throws \RedisClusterException
     */
    private function execute(callable $callback): mixed
    {
        $attempts = 0;
        $maxAttempts = 1 + $this->maxRetries;

        while ($attempts < $maxAttempts) {
            try {
                return $callback();
            } catch (\RedisClusterException $th) {
                if (! $this->isConnectionError($th)) {
                    throw $th;
                }

                $attempts++;

                if ($attempts >= $maxAttempts) {
                    throw $th;
                }

                usleep($this->retryDelay * 1000); // Convert milliseconds to microseconds

                try {
                    $this->reconnect();
                } catch (\RedisClusterException) {
                    // Reconnect failed, will retry on next iteration
                }
            }
        }

        // This line is unreachable but required for PHPStan
        throw new \RedisClusterException('Failed to execute Redis command');
    }

    /**
     * Check if the exception is a connection-related error that should trigger reconnect.
     *
     * RedisClusterException always returns error code 0 with no subclasses for different error types.
     * The only way to differentiate connection errors from command errors is by message matching.
     */
    private function isConnectionError(Exception $e): bool
    {
        $connectionErrors = [
            'went away',
            'socket',
            'read error on connection',
            'connection lost',
            'timed out',
            'timeout',
            'connection refused',
            'no connection',
            'broken pipe',
            // Redis Cluster specific
            "couldn't map cluster keyspace",
            "can't communicate with any node",
            'clusterdown',
            'is not covered by any node',
        ];

        $message = strtolower($e->getMessage());
        return array_any($connectionErrors, fn(string $needle): bool => str_contains($message, $needle));
    }

    private function reconnect(): void
    {
        $newRedis = new Client(
            $this->name,
            $this->seeds,
            $this->timeout,
            $this->readTimeout,
            $this->persistent,
            $this->auth,
        );

        $this->redis = $newRedis;
    }

    public function getName(?string $key = null): string
    {
        return 'redis-cluster';
    }
}
